### PR TITLE
HIVE-26431: Use correct schema for iceberg time travel queries

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
For time travel queries we currently always use the latest schema in IcebergSerDe, however since schemas are now versioned, we could use the schema which was active at the time in the past.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit test, manual test

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java [hunk 1]
```diff
@@ -105,8 +106,18 @@ public void initialize(@Nullable Configuration configuration, Properties serDePr
     } else {
       try {
         Table table = IcebergTableUtil.getTable(configuration, serDeProperties);
-        // always prefer the original table schema if there is one
-        this.tableSchema = table.schema();
+        if (serDeProperties.containsKey(serdeConstants.AS_OF_VERSION)) {
+          this.tableSchema = SnapshotUtil.schemaFor(table,
+              Long.valueOf(serDeProperties.getProperty(serdeConstants.AS_OF_VERSION)));
+          serDeProperties.remove(serdeConstants.AS_OF_VERSION);
```
Comment (szlta) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:112]: Let's explain in a comment why we need to remove this from serdeProperties, and perhaps also why we need serdeProperties to carry this information nstead of job configuration in the first place.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java [hunk 2]
```diff
@@ -347,7 +351,15 @@ final public Deserializer getDeserializer(boolean skipConfError) {
 
   final public Deserializer getDeserializerFromMetaStore(boolean skipConfError) {
     try {
-      return HiveMetaStoreUtils.getDeserializer(SessionState.getSessionConf(), tTable, metaTable, skipConfError);
+      String asOfTimestamp = null;
+      if (getAsOfTimestamp() != null) {
+        ZoneId zoneId = SessionState.get() == null ? new HiveConf().getLocalTimeZone() :
+            SessionState.get().getConf().getLocalTimeZone();
+        TimestampTZ time = TimestampTZUtil.parse(PlanUtils.stripQuotes(getAsOfTimestamp()), zoneId);
+        asOfTimestamp = Long.toString(time.toEpochMilli());
```
Comment (szlta) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java:359]: This calculation of Iceberg As Of information from string to long is found in multiple places, I think we should refactor this logic into a utility class of ql perhaps.

**Review by szlta — COMMENTED** [review]
Overall looks good, just small comments. Also could we add a test case for running a timetravel query on schema-evolved tables?

**Comment by InvisibleProgrammer** [comment]
@lcspinter , I wonder, is that PR abandoned? 

Can we take it over? 
Can you summarise what are the necessary next steps? 

Thank you,
Zsolt

**Pull request closed without merging**